rfcs/2020-06-18-2625-architecture-revisit.md

```diff
@@ -0,0 +1,504 @@
+reloading. This is a good time to reconsider our strategy for that functionality
+and decide if we could simplify while maintaining the most important user-facing
+benefits.
+
+#### Improved quality and consistency
+
+In the early days of Vector's development, we deliberately resisted introducing
+much overarching structure. With our limited knowledge of what Vector would
+become, we wanted to avoid introducing anything that would turn out to be [the
+wrong
+abstraction](https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction).
+
+At the current stage of the project, however, we have more than enough
+information to make informed decisions and the lack of structure is proving to
+be burden both for maintenance and onboarding new contributors. It's time to
+take what we've learned and put into place concrete patterns that make our code
+more consistent, less duplicative, and more strict in enforcing correct
+behavior.
+
+### Stretch goals
+
+In addition to the goals above, there are a few enhancements we should keep in
+mind while doing this design work. We don't necessarily seek to accomplish them
+as part of the same work, but we should help to lay the groundwork to make them
+possible and ensure we don't paint ourselves into a corner where they become
+overly difficult to accomplish later.
+
+#### Better branching
+
+We currently have a swimlanes transform that allows users to encode logical
+branching. This is a very valuable feature but had to be shoehorned into our
+existing topology model. Implementation-wise, it's not actually a transform at
+all, but a macro that expands to multiple transforms. While this is clever and
+useful, it is far from optimal in terms of complexity, performance, etc. With
+our work on the topology system, we should think about ways to support this more
+directly in the future.
+
+#### End-to-end acknowledgement
+
+For certain sources like Kafka, Vector's usefulness would be dramatically
+improved with the ability to precisely acknowledge messages once they have been
+fully processed. This would provide far more meaningful guarantees against data
+loss than we are able to provide today.
+
+We currently provide a way for sinks to propagate acknowledgements back to their
+disk buffer, but a real solution to this problem would need to span the entire
+pipeline. This also opens up questions about acknowledging messages that fan out
+to multiple sinks, etc.
+
+#### Multi-tenancy and hosted environments
+
+Vector was designed to be run continuously on a single host in a single
+organization's infrastructure. While this will like always be our primary use
+case, a few others have come up that would be beneficial to enable.
+
+One is multi-tenancy. What would it look like for a Vector process to support
+not just one topology, but many, across many organizations? This would likely
+push topologies to be more data-like and less physical collections of running
```

Comment:
Similar to [defunctionalization](https://blog.sigplan.org/2019/12/30/defunctionalization-everybody-does-it-nobody-talks-about-it/). We already do this to a degree with the way that Vector is configurable instead of a program that does one thing. This means to try to take that even further with our runtime representation and try to decouple topologies from things like running threads or tasks.